engine/graphics/src/null/graphics_null.cpp

```diff
@@ -735,217 +735,249 @@ namespace dmGraphics
         return g_DrawCount;
     }
 
-    struct ShaderBinding
-    {
-        ShaderBinding() : m_Name(0) {};
-        char*    m_Name;
-        uint32_t m_Index;
-        uint32_t m_Size;
-        uint32_t m_Stride;
-        Type     m_Type;
-    };
-
-    struct ShaderProgram
+    struct ResourceBindingDesc
     {
-        char*                  m_Data;
-        ShaderDesc::Language   m_Language;
-        dmArray<ShaderBinding> m_Uniforms;
-        dmArray<ShaderBinding> m_Attributes;
+        uint16_t m_Binding;
+        uint8_t  m_Taken;
     };
 
-    static void PushBinding(dmArray<ShaderBinding>* binding_array, const char* name, uint32_t name_length, dmGraphics::Type type, uint32_t size)
+    // TODO: This can be merged with a certain "blue" graphics adapter
+    static void FillProgramResourceBindings(
```

Comment:
We can perhaps merge this with all adapters as well (vulkan + webgpu has almost identical setup)